File: solutions/golang/pubsubsystem/pubsub_system_demo.go
package pubsubsystem

func Run() {
	// Create topics
	topic1 := NewTopic("Topic1")
	topic2 := NewTopic("Topic2")

	// Create publishers
	publisher1 := NewPublisher()
	publisher2 := NewPublisher()

	// Create subscribers
	subscriber1 := NewPrintSubscriber("Subscriber1")
	subscriber2 := NewPrintSubscriber("Subscriber2")
	subscriber3 := NewPrintSubscriber("Subscriber3")

	publisher1.RegisterTopic(topic1)
	publisher2.RegisterTopic(topic2)

	// Subscribe to topics
	topic1.AddSubscriber(subscriber1)
	topic1.AddSubscriber(subscriber2)
	topic2.AddSubscriber(subscriber2)
	topic2.AddSubscriber(subscriber3)

	// Publish messages
	publisher1.Publish(topic1, NewMessage("Message1 for Topic1"))
	publisher1.Publish(topic1, NewMessage("Message2 for Topic1"))
	publisher2.Publish(topic2, NewMessage("Message1 for Topic2"))

	// Unsubscribe from a topic
	topic1.RemoveSubscriber(subscriber2)

	// Publish more messages
	publisher1.Publish(topic1, NewMessage("Message3 for Topic1"))
	publisher2.Publish(topic2, NewMessage("Message2 for Topic2"))
}

File: solutions/java/src/pubsubsystem/PubSubSystemDemo.java
package pubsubsystem;

public class PubSubSystemDemo {
    public static void run() {
        // Create Broker
        Broker broker = new Broker();

        // Create Topics
        broker.createTopic("topic1");
        broker.createTopic("topic2");

        // Create publishers
        Publisher publisher1 = new Publisher("publisher1", broker);
        Publisher publisher2 = new Publisher("publisher2", broker);

        // Create subscribers
        Subscriber subscriber1 = new PrintSubscriber("PrintSubscriber1");
        Subscriber subscriber2 = new PrintSubscriber("PrintSubscriber2");
        Subscriber subscriber3 = new LoggingSubscriber("LoggingSubscriber3");

        // Subscribe to topics
        broker.subscribe("topic1", subscriber1);
        broker.subscribe("topic1", subscriber2);
        broker.subscribe("topic2", subscriber3);

        // Publish messages
        publisher1.publish("topic1", "Message1 for Topic1");
        publisher1.publish("topic1", "Message2 for Topic1");
        publisher1.publish("topic2", "Message1 for Topic2");

        // Unsubscribe from a topic
        broker.unsubscribe("topic1", subscriber2);

        // Publish more messages
        publisher1.publish("topic1", "Message3 for Topic1");
        publisher2.publish("topic2", "Message2 for Topic2");

        try {
            Thread.sleep(100); // Allow async delivery
            Dispatcher.shutdown();
        } catch (InterruptedException e) {
            System.out.println("Interrupted exception");
        }
    }
}
